### PR TITLE
Refactor VirtIO MMIO register map handling

### PR DESCRIPTION
**Pull request by shengwen-tw**
Two adjustments are conducted in this pull request:
1. Replace the hard-coded register address checking with enum.
2. Handle the configuration space read / write with device-specific structures.

### CODE REVIEW

File: virtio.h [hunk 1]
```diff
@@ -24,9 +24,82 @@
...
+    Status = 0x070 >> 2,            /* RW */
+    QueueDescLow = 0x080 >> 2,      /* W */
+    QueueDescHigh = 0x084 >> 2,     /* W */
+    QueueDriverLow = 0x090 >> 2,    /* W */
+    QueueDriverHigh = 0x094 >> 2,   /* W */
+    QueueDeviceLow = 0x0a0 >> 2,    /* W */
+    QueueDeviceHigh = 0x0a4 >> 2,   /* W */
+    ConfigGeneration = 0x0fc >> 2,  /* R */
+    Config = 0x100 >> 2,            /* RW */
+} VirtIO_MMIO_Reg_Map;
```
Comment (jserv) [inline virtio.h:53]: Avoid [Camel case](https://en.wikipedia.org/wiki/Camel_case).

**Comment by shengwen-tw** [comment]
The CI failure is due to the wrong syntax of the enum, i.e.,
```
enum NAME {...}; //correct syntax
enum {...}  NAME; //wrong syntax
```
Where the compilation passed on my Ubuntu 20.04 machine but failed on the 22.04 version's machine.

**Comment by jserv** [comment]
> The CI failure is due to the wrong syntax of the enum, i.e.,

You MUST use `clang-format-12`.

**Comment by jserv** [comment]
> ```
> enum NAME {...}; //correct syntax
> enum {...}  NAME; //wrong syntax
> ```

Use `typedef`.

File: virtio.h [hunk 2]
```diff
@@ -24,9 +24,82 @@
 #define VIRTIO_BLK_S_IOERR 1
 #define VIRTIO_BLK_S_UNSUPP 2
 
+/* VirtIO MMIO registers */
+enum {
+    MagicValue = 0x000 >> 2,        /* R */
```
[the remark this answers is not in the document]
Comment (shengwen-tw) [reply virtio.h]: Neat!

File: virtio.h [hunk 3]
```diff
@@ -24,9 +24,87 @@
...
+    _(QueueDeviceLow, 0x0a0)    /* W */  \
+    _(QueueDeviceHigh, 0x0a4)   /* W */  \
+    _(ConfigGeneration, 0x0fc)  /* R */  \
+    _(Config, 0x100)            /* RW */
+
+typedef enum {
+#define _(reg, addr) VIRTIO_##reg = addr >> 2,
+    VIRTIO_REG_LIST
+#undef _
+} VIRTIO_MMIO_REG_MAP;
```
Comment (jserv) [inline virtio.h:58]: Discard `VIRTIO_MMIO_REG_MAP` since its identifier does not matter.

**Review by jserv — CHANGES_REQUESTED** [review]
Move virtio-net and virtio-blk specific structure definitions from virtio.h to dedicated C source files.

Apply the trick [opaque pointer](https://en.wikipedia.org/wiki/Opaque_pointer) to hide the implementation details. That is, consider the following changes:
```diff
--- a/device.h
+++ b/device.h
@@ -144,14 +144,14 @@ typedef struct {
     /* queue config */
     uint32_t QueueSel;
     virtio_blk_queue_t queues[2];
-    /* device configuration */
-    struct virtio_blk_config config;
     /* status */
     uint32_t Status;
     uint32_t InterruptStatus;
     /* supplied by environment */
     uint32_t *ram;
     uint32_t *disk;
+    /* implementation-specific */
+    void *priv;
 } virtio_blk_state_t;
 
 void virtio_blk_read(vm_t *vm,
--- a/virtio-blk.c
+++ b/virtio-blk.c
@@ -21,6 +21,46 @@
 #define VBLK_QUEUE_NUM_MAX 1024
 #define VBLK_QUEUE (vblk->queues[vblk->QueueSel])
+
+#define PRIV(x) ((struct virtio_blk_config *) x->priv)
+
 static void virtio_blk_set_fail(virtio_blk_state_t *vblk)
 {
     vblk->Status |= VIRTIO_STATUS__DEVICE_NEEDS_RESET;
@@ -45,11 +85,11 @@ static void virtio_blk_update_status(virtio_blk_state_t *vblk, uint32_t status)
     /* Reset */
     uint32_t *ram = vblk->ram;
     uint32_t *disk = vblk->disk;
-    uint32_t capacity = vblk->config.capacity;
+    uint32_t capacity = PRIV(vblk)->capacity;
     memset(vblk, 0, sizeof(*vblk));
     vblk->ram = ram;
     vblk->disk = disk;
-    vblk->config.capacity = capacity;
+    PRIV(vblk)->capacity = capacity;
 }
 
 static void virtio_blk_write_handler(virtio_blk_state_t *vblk,
```

Of course, you have to allocate the necessary memory for the region pointed by `priv` member.

File: virtio-blk.c [hunk 4]
```diff
@@ -21,6 +21,40 @@
 #define VBLK_QUEUE_NUM_MAX 1024
 #define VBLK_QUEUE (vblk->queues[vblk->QueueSel])
 
+#define PRIV(x) ((struct virtio_blk_config *) x->priv)
+#define PRIV_INIT() malloc(sizeof(struct virtio_blk_config))
```
Comment (jserv) [inline virtio-blk.c:25]: Don't invoke `malloc`. Use static allocation instead. Be aware of the fact that semu does not use `malloc` so far. semu is designed for minimal memory consumption.

You can prepare a static global object which is an instance of `struct virtio_blk_config` and then point `priv` to it.

Comment (shengwen-tw) [reply virtio-blk.c:25]: That is fairly correct.

**Comment by jserv** [comment]
> Should we also add destructor functions to clean up the allocated memory (i.e., `vnet->priv` and `vblk->priv`) explicitly even if the program will only be terminated by pressing Ctrl+A then X?

Yes, you can do it in next pull request(s).
Follow the way how [kvm-host](https://github.com/sysprog21/kvm-host) does.

File: virtio-blk.c [hunk 5]
```diff
@@ -219,125 +255,139 @@ static bool virtio_blk_reg_read(virtio_blk_state_t *vblk,
...
         return true;
-    case 64: /* CapacityLow (R) */
-        *value = 0x00000000ffffffff & vblk->capacity;
-        return true;
-    case 65: /* CapacityHigh (R) */
-        *value = vblk->capacity >> 32;
-        return true;
     default:
-        return false;
+        /* Undefined register before the configuration space */
```
Comment (jserv) [inline virtio-blk.c:293]: You may introduce the `RANGE_CHECK` macro in `common.h`. See https://github.com/sysprog21/semu/blob/xv6/semu.c#L53
Then, you can rewrite as following:
```c
    if (!RANGE_CHECK(addr, _(Config), sizeof(struct virtio_blk_config)))
        return false;
```

File: virtio-blk.c [hunk 6]
```diff
@@ -219,125 +260,131 @@ static bool virtio_blk_reg_read(virtio_blk_state_t *vblk,
...
-        *value = vblk->capacity >> 32;
-        return true;
     default:
-        return false;
+        /* Invalid address which exceeded the range */
+        if (!RANGE_CHECK(addr, _(Config), sizeof(struct virtio_blk_config)))
+            return false;
+
+        /* Read configuration from the corresponding register */
+        uint32_t *config = (uint32_t *) PRIV(vblk);
```
Comment (jserv) [inline virtio-blk.c:303]: Simply rewrite as following:
```c
       *value = ((uint32_t *) PRIV(vblk))[addr - _(Config)];
```

This change avoids the mixture of `config` (lowercase) and `Config` (uppercase), which might confuse.

File: virtio-net.c [hunk 7]
```diff
@@ -218,109 +235,121 @@ static bool virtio_net_reg_read(virtio_net_state_t *vnet,
...
+     * touches the MAC address of the virtio-net under 8-bit accesses
      */
     default:
-        return false;
+        /* Invalid address which exceeded the range */
+        if (!RANGE_CHECK(addr, _(Config), sizeof(struct virtio_net_config)))
+            return false;
+
+        /* Read configuration from the corresponding register */
+        uint32_t *config = (uint32_t *) PRIV(vnet);
```
Comment (jserv) [inline virtio-net.c:286]: Ditto. Avoid the variable `config`.

**Review by jserv — CHANGES_REQUESTED** [review]
Rebase the latest `master` branch and resolve the conflicts.

File: virtio-blk.c [hunk 8]
```diff
@@ -386,11 +431,20 @@ void virtio_blk_write(vm_t *vm,
...
 uint32_t *virtio_blk_init(virtio_blk_state_t *vblk, char *disk_file)
 {
+    if (vblk_dev_cnt >= VBLK_DEV_CNT_MAX) {
+        fprintf(stderr,
+                "Excedded the number of virtio-blk device can be allocated.\n");
+        exit(2);
+    }
+
+    /* Allocate memory for the private member */
+    vblk->priv = (void *) &vblk_configs[vblk_dev_cnt++];
```
Comment (jserv) [inline virtio-blk.c:441]: The cast to `void *` is not necessary as C standard allows implicit casting to `void *`.

File: virtio-net.c [hunk 9]
```diff
@@ -395,6 +431,15 @@ void virtio_net_write(vm_t *vm,
...
 bool virtio_net_init(virtio_net_state_t *vnet)
 {
+    if (vnet_dev_cnt >= VNET_DEV_CNT_MAX) {
+        fprintf(stderr,
+                "Excedded the number of virtio-net device can be allocated.\n");
+        exit(2);
+    }
+
+    /* Allocate memory for the private member */
+    vnet->priv = (void *) &vnet_configs[vnet_dev_cnt++];
```
Comment (jserv) [inline virtio-net.c:441]: Ditto. Remove unnecessary casting.

**Review by jserv — CHANGES_REQUESTED** [review]
Improve the git commit message. This commit addresses the following:
1. Re-organizes the source structures to improve code organization and maintainability.
2. Introduces the use of [opaque pointer](https://en.wikipedia.org/wiki/Opaque_pointer for information hiding, enhancing encapsulation and abstraction.
3. Enforces mnemonic enums instead of numbers, enhancing code readability and maintainability.

By incorporating these changes, the codebase becomes more structured, secure, and easier to understand and maintain."

**Review by jserv — CHANGES_REQUESTED** [review]
Rebase the latest `master` branch.

**Comment by jserv** [comment]
Thank @shengwen-tw for contributing!

**Pull request merged**